mitodl/mitxonline
# Adding Discount UI

#### Pre-Flight checklist


#### What are the relevant tickets?
Fixes #431

#### What's this PR do?
Adds support for discount redemption:
* Adds UI to apply and clear discount codes
* Updates backend APIs to handle discount code redemption following the established rules
* Updates the Django Admin to include screens for discount redemptions, and to display discounts applied on each order.

#### How should this be manually tested?
Ensure your setup has a valid product and course combination set up. 
Add a discount code in via djadmin and note the code you create.
Add a product to your cart, and then apply the discount code using the form in Order Summary.
The discount should be applied successfully and the price should reflect the discounted amount.

To test adding multiple codes to an order: follow the same instructions as above but add more than one discount. Then, apply a discount code after applying a discount code. The new code should take precedence over the old one.

To test user discounts: follow the first two steps under the standard test, then also add a User Discount record. Log in as the user you added the discount for and add a product to the cart. You should see that the discount has been applied to the cart automatically. Applying another discount over the user discount should override the user discount, but clearing the discount should re-apply the user discount. Other accounts in the system should not have the discount applied to their carts automatically. 

#### Any background context you want to provide?
Rules as follows:
- Only a single discount can be applied to a given cart - any additional discounts added after the first override the last added discount. 
- User discounts are always applied, unless overridden by another discount.
- Discounts that have hit the maximum number of redemptions (either by hitting the value in max_redemptions, or by virtue of being a "one-time" or "one-time-per-user" discount type) will return a "not found" and won't be applied. 
- Discounts are applied to the total amount in the cart (there's only one allowed item in the cart right now, though)

Also of note:
- Error and success states right now pop a standard banner notification. This wasn't in the InVision mockup - seemed reasonable so included it. 
- The "Adding another discount" text disappears if there's not already a discount applied.
- Clear Discount only appears if there's a discount to clear. The front-end has no idea (right now) if a discount is a user discount or not so you can clear a user discount; it just won't apparently do anything as the discount will just get reapplied. 
- Fixed Price discounts probably need some styling help - I added Fixed Price text in there to differentiate them from the dollar off or percent off discounts. Input appreciated for this. 

#### Screenshots (if appropriate)
Desktop:
![Screen Shot 2022-02-25 at 2 25 00 PM](https://user-images.githubusercontent.com/945611/155799877-16eb64ee-44dc-4856-838a-0e5e6571cc98.png)
![Screen Shot 2022-02-25 at 2 25 34 PM](https://user-images.githubusercontent.com/945611/155799880-54b2f09e-1cfe-4b3c-aad6-4d8ffb1651e6.png)
![Screen Shot 2022-02-25 at 2 26 07 PM](https://user-images.githubusercontent.com/945611/155799882-8c7dc97c-0a84-46e0-a8c9-118561b97136.png)

Mobile:
![Screen Shot 2022-02-25 at 2 25 18 PM](https://user-images.githubusercontent.com/945611/155799879-d4a8d172-ab51-4cea-b521-a139bf5e1898.png)
![Screen Shot 2022-02-25 at 2 25 27 PM](https://user-images.githubusercontent.com/945611/155799905-5f856f7e-5bc5-46dd-80c7-1ade5750d8ee.png)
![Screen Shot 2022-02-25 at 2 26 11 PM](https://user-images.githubusercontent.com/945611/155799907-c1aafb17-f6ec-4189-ae15-5abc9d90c6f7.png)
